cloud/scope/powervs_cluster_test.go

```diff
@@ -3043,3 +3043,1476 @@ func TestCreateVPCSubnet(t *testing.T) {
+				return &clusterScope
+			},
+			expectedErr: true,
+		},
+		{
+			name: "When Security group rule is created successfully but returns error while creating security group rules",
+			prepareParamsFunc: func(mockVPC *mock.MockVpc) *PowerVSClusterScope {
+				remote := infrav1beta2.VPCSecurityGroupRuleRemote{
+					Address:    ptr.To("192.168.0.1/24"),
+					RemoteType: infrav1beta2.VPCSecurityGroupRuleRemoteTypeAddress,
+				}
+
+				rules := infrav1beta2.VPCSecurityGroupRule{
+					Direction: infrav1beta2.VPCSecurityGroupRuleDirectionInbound,
+					Source: &infrav1beta2.VPCSecurityGroupRulePrototype{
+						Remotes:  append([]infrav1beta2.VPCSecurityGroupRuleRemote{}, remote),
+						Protocol: infrav1beta2.VPCSecurityGroupRuleProtocolTCP,
+						PortRange: &infrav1beta2.VPCSecurityGroupPortRange{MaximumPort: 65535,
+							MinimumPort: 1,
+						},
+					},
+				}
+
+				clusterScope := PowerVSClusterScope{
+					IBMVPCClient: mockVPC,
+					IBMPowerVSCluster: &infrav1beta2.IBMPowerVSCluster{
+						Status: infrav1beta2.IBMPowerVSClusterStatus{
+							VPC: &infrav1beta2.ResourceReference{
+								ID: ptr.To("securityGroupID"),
+							},
+						},
+						Spec: infrav1beta2.IBMPowerVSClusterSpec{
+							VPCSecurityGroups: append([]infrav1beta2.VPCSecurityGroup{}, infrav1beta2.VPCSecurityGroup{
+								Name:  ptr.To("securityGroupName"),
+								Rules: append([]*infrav1beta2.VPCSecurityGroupRule{}, &rules),
+							}),
+						},
+					},
+				}
+
+				mockVPC.EXPECT().GetSecurityGroupByName(gomock.Any()).Return(nil, nil)
+				mockVPC.EXPECT().CreateSecurityGroup(gomock.Any()).Return(&vpcv1.SecurityGroup{ID: ptr.To("securityGroupID")}, nil, nil)
+				mockVPC.EXPECT().CreateSecurityGroupRule(gomock.Any()).Return(nil, nil, errors.New("failed to create security group rule"))
+				return &clusterScope
+			},
+			expectedErr: true,
+		},
+	}
+
+	for _, tc := range testCases {
+		g := NewWithT(t)
+		setup(t)
+		t.Cleanup(teardown)
+		t.Run(tc.name, func(_ *testing.T) {
+			clusterScope := tc.prepareParamsFunc(mockVPC)
+			err := clusterScope.ReconcileVPCSecurityGroups()
+			if tc.expectedErr {
+				g.Expect(err).ToNot(BeNil())
+			} else {
+				g.Expect(err).To(BeNil())
+			}
+		})
+	}
+}
+
+func TestValidateVPCSecurityGroup(t *testing.T) {
+	var (
+		mockVPC  *mock.MockVpc
+		mockCtrl *gomock.Controller
+	)
+
+	setup := func(t *testing.T) {
+		t.Helper()
+		mockCtrl = gomock.NewController(t)
+		mockVPC = mock.NewMockVpc(mockCtrl)
+	}
+	teardown := func() {
+		mockCtrl.Finish()
+	}
+
+	type response struct {
+		rulesIDs []*string
+		sg       *vpcv1.SecurityGroup
+	}
+
+	testCases := []struct {
+		name              string
+		expectedErr       bool
+		expectedResponse  response
+		prepareParamsFunc func(*mock.MockVpc) (*PowerVSClusterScope, infrav1beta2.VPCSecurityGroup)
+	}{
+		{
+			name: "When GetVPCSubnetByName returns error while comparing security group rules",
```

Comment:
Why are you doing this at this scope? This case actually happening at `validateVPCSecurityGroupRuleRemote` its after few func calls from this func. 
Since you have added separate UT for all the funcs it's better to limit the UT specific to the func you are testing.
If we do like this it would be confusing and we might miss some cases.